rfcs/message-first-architecture/RFC.md

```diff
@@ -0,0 +1,211 @@
+- Complicated application logic because apps care about messages, not resources.
+- Potentially, technical debt when extending inlang (core) beyong messages to other types of content (e.g. markdown, images, etc.).
+
+### Wrong assumptions with core's "resource-first" architecture
+
+#### 1. [False] Resources are the primary data structure of inlang
+
+That's not true. Messages is what matters to apps and users.
+
+No inlang application does anything with resources. However, plugins do. The derived insight is that plugins deal with resources because plugins are the interface between app and source code. If this insight is true, apps should not care about resources.
+
+#### 2. [False] A resource AST is required
+
+Turned out to be false. Plugins re-construct resources from Messages, not from a resource AST.
+
+Resources are almost always key-value data structures with a parsing and serialization spec. Precise knowledge how a resource file is "construcuted" e.g. message id "5" starts on line 3 column 5 is not required. As a matter of fact, the AST approach for resources doesn't even work because CRUD operations, by nature, do not contain information like message id "5" starts on line 3 column 5. CRUD operations not providing AST information lead to the [removal of `metadata`](https://github.com/inlang/inlang/issues/945).
+
+## Proposal
+
+Inlang core should be refactored to be "message-first".
+
+- Simplifies the application logic of inlang core and inlang apps.
+- Simplify communication of inlang because the concept of "resources" is abstracted away/only plugins care about resources.
+- Future-proofs inlang. Messages can be retrieved from anywhere (e.g. resources files, remote APIs, embedded databases).
+
+### Application flow
+
+> Don't take the API design too seriously. It's a sketch to illustrate the simplification.
+
+- The general application flow is vastly simplified: No more resources, query, etc. Just `messages`.
+- The `messages` object can be performance optimized internally.
+
+#### 1. `config.getMessages()`
+
+- Plugins can retrieve messages from anywhere (e.g. resources files, remote APIs, embedded databases).
+- Future optimization: Only retrieve messages that are actually used in the app by passing a filter object.
+
+```ts
+const messages = await config.getMessages()
+```
+
+#### 2. CRUD operations
+
+- No more `resources` or `query` objects
+- Way simpler API.
+
+```ts
+messages = messages.create({
+	id: "hello",
+	languageTag: "en-US",
+	body: "how are you?",
+})
+
+const message = messages.get({
+	id: "hello",
+	languageTag: "en-US",
+	body: "how are you?",
+})
+
+messages = messages.update({
+	id: "hello",
+	languageTag: "en-US",
+	body: "how are you doing?",
+})
+
+messages = messages.delete({
+	id: "hello",
+	languageTag: "en-US",
+})
+```
+
+```ts
+// For comparison, here is the current API:
+
+const refResource = resources.find((resource) => resource.languageTag.name === "en-US")
+const message = query(refResource).get({
```

Comment:
To be fair this API could also be simplified to the example below ;)

Reply:
True but with more effort. messages would need to deal with resources under the hood and yikes everything becomes more complicated :D